src/Uno.Toolkit.UI/Controls/AutoLayout/AutoLayout.Layouting.Arrange.cs

```diff
@@ -304,28 +316,40 @@ private static void ApplyMinMaxValues(UIElement element, Orientation orientation
 
 	[MethodImpl(MethodImplOptions.AggressiveInlining)]
 	private static double PrimaryAxisAlignmentOffsetSize(
+		double totalOfFillMaxSize,
 		AutoLayoutAlignment autoLayoutAlignment,
 		double availableSizeForStackedChildren,
 		double totalNonFilledStackedSize,
-		bool atLeastOneFilledChild,
+		bool haveMoreFilled,
 		double spacing,
 		int numberOfStackedChildren)
 	{
-		if (atLeastOneFilledChild || autoLayoutAlignment is AutoLayoutAlignment.Start)
+		if (haveMoreFilled || autoLayoutAlignment is AutoLayoutAlignment.Start)
 		{
 			return 0;
 		}
 
 		var totalSpacingSize = spacing * (numberOfStackedChildren - 1);
-
 		var alignmentOffsetSize = availableSizeForStackedChildren -
-			(totalNonFilledStackedSize + totalSpacingSize);
+			(totalNonFilledStackedSize + totalSpacingSize //);
+			+ totalOfFillMaxSize);
 
 		return autoLayoutAlignment switch
 		{
-			AutoLayoutAlignment.End => alignmentOffsetSize,
+			AutoLayoutAlignment.End when alignmentOffsetSize > 0 => alignmentOffsetSize,
 			AutoLayoutAlignment.Center when alignmentOffsetSize > 0 => alignmentOffsetSize / 2,
 			_ => 0
 		};
 	}
+
+	[MethodImpl(MethodImplOptions.AggressiveInlining)]
+	private static double GetChildrenLowerThanAllocateSurplus(CalculatedChildren[] calculatedChildren, double filledChildrenSizeBeforeMaxSize)
+	{
+		var filledChildrenGreaterThanAllocated = calculatedChildren.Where(c => c.Role is AutoLayoutRole.Filled && c.MeasuredLength > filledChildrenSizeBeforeMaxSize);
+		var filledChildrenLowerThanAllocated = calculatedChildren.Where(c => c.Role is AutoLayoutRole.Filled && c.MeasuredLength <= filledChildrenSizeBeforeMaxSize);
+
+		var surplusFromLower = filledChildrenLowerThanAllocated.Select(c => filledChildrenSizeBeforeMaxSize - c.MeasuredLength);
+
+		return filledChildrenGreaterThanAllocated.Any() ? surplusFromLower.Sum() / filledChildrenGreaterThanAllocated.Count() : 0;
```

Comment:
Please remove the Linq usage and make only one loop. You are iterating the children more than once and we shouldn't do that in a such intensive panel :-)

You code is pretty and acceptable in most situations, but not here.

Try to make a for loop instead of a foreach to minimize memory allocation.

Reply:
```suggestion
		int countFilledChildrenGreaterThanAllocated = 0;
		double sumSurplusFromLower = 0;

		for (int i = 0; i < calculatedChildren.Length; i++)
		{
			if (calculatedChildren[i].Role is AutoLayoutRole.Filled)
			{
				if (calculatedChildren[i].MeasuredLength > filledChildrenSizeBeforeMaxSize)
				{
					countFilledChildrenGreaterThanAllocated++;
				}
				else
				{
					sumSurplusFromLower += filledChildrenSizeBeforeMaxSize - calculatedChildren[i].MeasuredLength;
				}
			}
		}

		return countFilledChildrenGreaterThanAllocated > 0 ? sumSurplusFromLower / countFilledChildrenGreaterThanAllocated : 0;
```